index.js

```diff
@@ -207,56 +207,90 @@ class ServerlessCustomDomain {
+        }
+      });
+    }
+
+    if (certificateArn == null) {
+      throw Error(`Could not find the certificate ${certificateName}`);
+    }
+    return certificateArn;
+  }
+
+  /*
    * Obtains the certification arn
    */
   getCertArn() {
     const acm = new AWS.ACM({
       region: 'us-east-1',
     });       // us-east-1 is the only region that can be accepted (3/21)
 
-    const certArn = acm.listCertificates().promise();
-
-    return certArn.then((data) => {
-      // The more specific name will be the longest
-      let nameLength = 0;
-      // The arn of the choosen certificate
-      let certificateArn;
-      // The certificate name
-      let certificateName = this.serverless.service.custom.customDomain.certificateName;
+    const issuedCertArn = acm.listCertificates({ CertificateStatuses: ['ISSUED'] }).promise();
 
+    return issuedCertArn.then((data) => {
+      if (process.env.SLS_DEBUG && Array.isArray(data.CertificateSummaryList)) {
+        this.serverless.cli.log(`Found ${data.CertificateSummaryList.length} Valid Certificates: ${JSON.stringify(data.CertificateSummaryList)}`);
+      }
+      return this.findClosestCertificate(data);
+    }).catch((err) => {
+      // No cert found in the ISSUED status. Let's look through certs in all other statuses.
+      const certArn = acm.listCertificates().promise();
 
-      // Checks if a certificate name is given
-      if (certificateName != null) {
-        const foundCertificate = data.CertificateSummaryList
-          .find(certificate => (certificate.DomainName === certificateName));
-
-        if (foundCertificate != null) {
-          certificateArn = foundCertificate.CertificateArn;
+      return certArn.then((data) => {
+        if (process.env.SLS_DEBUG && Array.isArray(data.CertificateSummaryList)) {
+          this.serverless.cli.log(`Found ${data.CertificateSummaryList.length} Certificates: ${JSON.stringify(data.CertificateSummaryList)}`);
         }
-      } else {
-        certificateName = this.givenDomainName;
-        data.CertificateSummaryList.forEach((certificate) => {
-          let certificateListName = certificate.DomainName;
-
-          // Looks for wild card and takes it out when checking
-          if (certificateListName[0] === '*') {
-            certificateListName = certificateListName.substr(1);
-          }
-
-          // Looks to see if the name in the list is within the given domain
-          // Also checks if the name is more specific than previous ones
-          if (certificateName.includes(certificateListName)
-            && certificateListName.length > nameLength) {
-            nameLength = certificateListName.length;
-            certificateArn = certificate.CertificateArn;
-          }
-        });
-      }
 
-      if (certificateArn == null) {
-        throw Error(`Could not find the certificate ${certificateName}`);
-      }
-      return certificateArn;
+        return this.findClosestCertificate(data);
```

Comment:
Does returning anything here actually do anything? Is this only here so that it can throw an error if none of the certs match. In that case maybe we can remove the `return` and leave a comment explaining what this is doing

Reply:
Yes, it is a nested promise inside of the catch.  It is used to continue to process the error pipeline and used on line 280. If it is not returned the error handling would stop and it wouldn't be able to get the details of the cert. See the diagram below:

![diagram](https://go.gliffy.com/go/share/image/s7tcd12w0orfhl5qds4y.png?utm_medium=live-embed&utm_source=custom)

Reply:
Also if you attempt to use a not issued cert pretty much anywhere the AWS SDK is going to throw an exception. This is because the AWS SDK methods that work with certs are expecting the cert to be in the issued status. 

This is what we ran into when attempting to this plugin initially. Since the error is not detected early a different error is thrown in the next step when trying to assign the cert to the custom domain. Hunting down the root cause was difficult because the error didn't explain that the cert was expired instead it looked like creating the domain was failing.

The objective of this pull request is to make sure the plugin will output a very specific error, hence all of the code being inside of the "catch" block. The specific error will allow the user to take action and correct it. Otherwise it just looks like the plugin doesn't work when really the cert was simply not in a valid status.

Reply:
Hey @exocom -- thanks so much for all this great work. I agree that the detail level and ability to accurately pin down issues has not been a strong suit of this plugin and it's great that you've been able to add that as I'm sure it'll save others a great deal of confusion while debugging. That said, is there any way we could break this up and make it easier to follow for people who may not be the world's best ES6 programmers? I don't doubt that it accurately identifies the problem, but if the code to reach a detailed error message is as cryptic as the original error message was, then in a sense we've taken one step forward and another step back.

Reply:
Hey @majones-amplify that is a hard question to answer. I appreciate simplicity in code just as much as the next person, however, I would take into consideration of how often does this code change and also does it have test coverage.

With code that has good test coverage and that rarely changes my team tends to just move on. Basically if the existing code uses callback then we contribute using callback, with promises we write promises, and when it is async await we write it with async await. The intent being that if you are using promises then the people debugging should be familiar with them. If not that is really on the developer to learn more about it. Which for us is how we've learned a ton from code on github, because most people do not hold back.

With that said, Ideally we would write this with ES2017 and use async await if the goal was extensibility. That means a complete rewrite, not recommended unless your team is really comfortable with it already.